File: workstation/src/main/java/io/xj/model/util/CsvUtils.java
// Copyright (c) XJ Music Inc. (https://xj.io) All Rights Reserved.
package io.xj.model.util;

import jakarta.annotation.Nullable;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.UUID;
import java.util.stream.Collectors;

public interface CsvUtils {
  String COMMA = ", ";
  char SPACE = ' ';

  static List<String> split(@Nullable String csv) {
    if (StringUtils.isNullOrEmpty(csv)) return List.of();
    return Arrays.stream(Objects.requireNonNull(csv).split(","))
      .map(String::trim)
      .collect(Collectors.toList());
  }

  static List<String> splitProperSlug(String csv) {
    Collection<String> items = split(csv);
    List<String> slugs = new ArrayList<>();
    items.stream().filter(item -> Objects.nonNull(item) && !item.isEmpty()).map(StringUtils::toProperSlug).forEach(slugs::add);
    return slugs;
  }

  static String join(List<?> parts) {
    return parts.stream().map(Object::toString).collect(Collectors.joining(COMMA));
  }

  @SafeVarargs
  static <E extends Enum<E>> String joinEnum(E... objects) {
    List<String> objectStrings = Arrays.stream(objects).filter(Objects::nonNull).map(Enum::toString).collect(Collectors.toList());
    return join(objectStrings);
  }

  /**
   Join a set of items' toString() values properly, e.g. "One, Two, Three, and Four"

   @param ids             to write
   @param beforeFinalItem text after last comma
   @return CSV of ids
   */
  static <T> String prettyFrom(Collection<T> ids, String beforeFinalItem) {
    if (Objects.isNull(ids) || ids.isEmpty()) {
      return "";
    }
    Iterator<T> it = ids.iterator();
    StringBuilder result = new StringBuilder(it.next().toString());
    while (it.hasNext()) {
      result.append(COMMA);
      String item = it.next().toString();
      if (!it.hasNext())
        result.append(beforeFinalItem).append(SPACE);
      result.append(item);
    }
    return result.toString();
  }

  /**
   Get a CSV string of key=value properties

   @param properties key=value
   @return CSV string
   */
  static String from(Map<String, String> properties) {
    List<String> pieces = new ArrayList<>();
    properties.forEach((key, value) -> pieces.add(String.format("%s=%s", key, value)));
    return join(pieces);
  }

  /**
   Get a CSV string of key=value properties

   @param properties key=value
   @return CSV string
   */
  static String from(Collection<?> properties) {
    return join(properties.stream().map(Objects::toString).collect(Collectors.toList()));
  }

  /**
   Split a string into a list of UUIDs

   @param uuids string to split
   @return list of UUIDs
   */
  static List<UUID> splitUUIDs(String uuids) {
    return split(uuids).stream().map(UUID::fromString).collect(Collectors.toList());
  }
}

File: engine/src/util/CsvUtils.cpp
// Copyright (c) XJ Music Inc. (https://xj.io) All Rights Reserved.

#include "xjmusic/util/CsvUtils.h"

using namespace XJ;


std::vector<std::string> CsvUtils::split(const std::string &csv) {
  if (csv.empty()) return {};
  std::vector<std::string> result;
  std::stringstream ss(csv);
  std::string item;
  while (std::getline(ss, item, ',')) {
    item.erase(item.begin(), std::find_if(item.begin(), item.end(), [](const unsigned char ch) {
                 return !std::isspace(ch);
               }));
    item.erase(std::find_if(item.rbegin(), item.rend(), [](const unsigned char ch) {
                 return !std::isspace(ch);
               }).base(),
               item.end());
    result.push_back(item);
  }
  return result;
}


std::vector<std::string> CsvUtils::splitProperSlug(const std::string &csv) {
  std::vector<std::string> items = split(csv);
  std::vector<std::string> slugs;
  for (const auto &item: items) {
    if (!item.empty()) {
      slugs.push_back(StringUtils::toProperSlug(item));
    }
  }
  return slugs;
}


std::string CsvUtils::join(const std::vector<std::string> &parts) {
  return StringUtils::join(parts, ", ");
}


std::string CsvUtils::prettyFrom(const std::vector<std::string> &ids, const std::string &beforeFinalItem) {
  if (ids.empty()) {
    return "";
  }
  std::ostringstream result;
  for (size_t i = 0; i < ids.size(); ++i) {
    if (i != 0) {
      result << ", ";
    }
    if (i == ids.size() - 1) {
      result << beforeFinalItem << ' ';
    }
    result << ids[i];
  }
  return result.str();
}


std::string CsvUtils::from(const std::map<std::string, std::string> &properties) {
  int i = 0;
  std::string result;
  for (const auto &[key, value]: properties) {
    if (i != 0) {
      result += ", ";
    }
    result += key;
    result += "=";
    result += value;
    i++;
  }
  return result;
}

std::string CsvUtils::toProperCsvAnd(const std::vector<std::string> &items) {
  return toProperCsv(items, "and");
}

std::string CsvUtils::toProperCsvOr(const std::vector<std::string> &items) {
  return toProperCsv(items, "or");
}

std::string CsvUtils::toProperCsv(std::vector<std::string> items, const std::string &finalSeparator) {
  if (items.empty()) return "";
  if (items.size() == 1) return StringUtils::toProper(items.at(0));
  if (items.size() == 2)
    return StringUtils::toProper(items.at(0)) + " " + finalSeparator + " " + StringUtils::toProper(items.at(1));
  auto allButLast = std::vector(items.begin(), items.end() - 1);
  return join(allButLast) + ", " + finalSeparator + " " + StringUtils::toProper(*items.end());
}
